File: Exercism/c/queen-attack/src/queen_attack.c
#include "queen_attack.h"
#include <stdlib.h>
#include <stdbool.h>

#define is_position_valid(x) ((x.row >= 0 && x.row < 8) && (x.column >= 0 && x.column < 8))

attack_status_t can_attack(position_t queen_1, position_t queen_2) {
    // Check if both queens are on valid positions
    if (!(is_position_valid(queen_1) && is_position_valid(queen_2)))
        return INVALID_POSITION;

    // Check if queens have different positions from each other
    if (queen_1.row == queen_2.row && queen_1.column == queen_2.column)
        return INVALID_POSITION;

    // Can attack on row or column
    if (queen_1.row == queen_2.row || queen_1.column == queen_2.column)
        return CAN_ATTACK;

    // Can attack on diagonal
    if (abs(queen_1.row-queen_2.row) == abs(queen_1.column-queen_2.column))
        return CAN_ATTACK;

    return CAN_NOT_ATTACK;
}

File: Exercism/python/queen-attack/queen_attack.py
class Queen:
    def __init__(self, row, column):
        if row < 0 or column < 0 or row > 7 or column > 7:
            raise ValueError("Invalid position")

        self.row = row
        self.column = column

    def can_attack(self, other):
        if self.row == other.row and self.column == other.column:
            raise ValueError("Queens can't have the same position")

        return self.row == other.row or self.column == other.column \
            or abs(self.row - other.row) == abs(self.column - other.column)